WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingCategoriesViewModel.kt

```diff
@@ -0,0 +1,223 @@
+import androidx.lifecycle.MediatorLiveData
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.launch
+import org.wordpress.android.R.string
+import org.wordpress.android.analytics.AnalyticsTracker.Stat
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.store.TaxonomyStore.OnTermUploaded
+import org.wordpress.android.models.CategoryNode
+import org.wordpress.android.modules.BG_THREAD
+import org.wordpress.android.ui.pages.SnackbarMessageHolder
+import org.wordpress.android.ui.posts.PrepublishingCategoriesViewModel.SwipeToRefreshUiState.NotRefreshingUiState
+import org.wordpress.android.ui.posts.PrepublishingCategoriesViewModel.SwipeToRefreshUiState.RefreshingUiState
+import org.wordpress.android.ui.posts.PrepublishingCategoriesViewModel.UiState.ContentUiState
+import org.wordpress.android.ui.utils.UiString
+import org.wordpress.android.ui.utils.UiString.UiStringRes
+import org.wordpress.android.util.NetworkUtilsWrapper
+import org.wordpress.android.util.analytics.AnalyticsTrackerWrapper
+import org.wordpress.android.viewmodel.Event
+import org.wordpress.android.viewmodel.ScopedViewModel
+import javax.inject.Inject
+import javax.inject.Named
+
+class PrepublishingCategoriesViewModel @Inject constructor(
+    private val getCategoriesUseCase: GetCategoriesUseCase,
+    private val updatePostCategoriesUseCase: UpdatePostCategoriesUseCase,
+    private val analyticsTrackerWrapper: AnalyticsTrackerWrapper,
+    private val networkUtilsWrapper: NetworkUtilsWrapper,
+    @Named(BG_THREAD) private val bgDispatcher: CoroutineDispatcher
+) : ScopedViewModel(bgDispatcher) {
+    private var isStarted = false
+    private lateinit var editPostRepository: EditPostRepository
+    private lateinit var siteModel: SiteModel
+    private var updateCategoriesJob: Job? = null
+
+    private val _navigateToHomeScreen = MutableLiveData<Event<Unit>>()
+    val navigateToHomeScreen: LiveData<Event<Unit>> = _navigateToHomeScreen
+
+    private val _navigateToAddCategoryScreen = MutableLiveData<Event<Unit>>()
+    val navigateToAddCategoryScreen: LiveData<Event<Unit>> = _navigateToAddCategoryScreen
+
+    private val _toolbarTitleUiState = MutableLiveData<UiString>()
+    val toolbarTitleUiState: LiveData<UiString> = _toolbarTitleUiState
+
+    private val _snackbarEvents = MediatorLiveData<Event<SnackbarMessageHolder>>()
+    val snackbarEvents: LiveData<Event<SnackbarMessageHolder>> = _snackbarEvents
+
+    private val _uiState: MutableLiveData<UiState> = MutableLiveData()
+    val uiState: LiveData<UiState> = _uiState
+
+    private val _refreshingUiState = MutableLiveData<SwipeToRefreshUiState>()
+    val refreshingUiState: LiveData<SwipeToRefreshUiState> = _refreshingUiState
+
+    fun start(editPostRepository: EditPostRepository, siteModel: SiteModel) {
+        this.editPostRepository = editPostRepository
```

Comment:
Noted in #13066 under arch changes.

Reply:
Actually @malinajirka I think those are fine - hence the use of the `this` keyword. They are being initialized not reassigned. I've seen this pattern in [Interests](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/reader/discover/interests/ReaderInterestsViewModel.kt#L53), [Stories](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/stories/StoryComposerViewModel.kt#L64), [PostList](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/posts/PostListMainViewModel.kt#L227), [Tags](https://github.com/wordpress-mobile/WordPress-Android/blob/2476981600502ddb9a8c73207e937fc1edb48707/WordPress/src/main/java/org/wordpress/android/ui/posts/PrepublishingTagsViewModel.kt#L39)

I mean, we could all be wrong 🤣  The other option is setting them var xxx = null? Thoughts?

Reply:
Isn't the `this` keyword used because of the name collision - both the parameter and the field are called `editPostRepository`?

> They are being initialized not reassigned.

But start() can be called multiple times, right? So the first call is "initialization" but any subsequent calls are "reassignments".

Either way, I've created a unit test and it seems "lateinit vars" can be re-assigned. Which kind of make sense, otherwise, they'd be called "lateinit val". So we are good - even though it's probably not a great practice since we are opening doors for race conditions. Having said that it's completely fine to use it here - we already use it at other places as you mentioned.

Reply:
Ah, yes, I forgot about `start` being called over and over again. It does seem like a poor pattern to rely on, I'll drop it on my list for hack week when I consolidate editPostSettings in. Thanks